File: server/src/controllers/roleController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import { v4 as uuidv4 } from 'uuid';
import { AuthRequest } from '../middleware/auth.js';

// Get all roles with permissions
export const getRoles = async (req: Request, res: Response) => {
  try {
    const { search, sortBy = 'name', sortOrder = 'asc' } = req.query;

    // Build WHERE clause for search
    let whereClause = '';
    const queryParams: any[] = [];

    if (search && typeof search === 'string' && search.trim()) {
      whereClause = 'WHERE name LIKE $1 OR description LIKE $2';
      const searchTerm = `%${search.trim()}%`;
      queryParams.push(searchTerm, searchTerm);
    }

    // Validate sortBy to prevent SQL injection
    const allowedSortFields = ['name', 'description', 'created_at', 'updated_at'];
    const validSortBy = allowedSortFields.includes(sortBy as string) ? sortBy : 'name';
    const validSortOrder = sortOrder === 'desc' ? 'DESC' : 'ASC';

    // Build and execute query
    const sql = `SELECT * FROM roles ${whereClause} ORDER BY ${validSortBy} ${validSortOrder}`.trim();
    const roles = await query<any[]>(sql, queryParams);

    // Get permissions for each role
    const rolesWithPermissions = await Promise.all(
      roles.map(async (role) => {
        const permissions = await query<any[]>(
          'SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?',
          [role.id]
        );
        return {
          ...role,
          permissions: permissions.reduce((acc, perm) => {
            acc[perm.permission_type] = perm.allowed;
            return acc;
          }, {} as Record<string, boolean>),
        };
      })
    );

    res.json(rolesWithPermissions);
  } catch (error: any) {
    console.error('Error fetching roles:', error);
    res.status(500).json({ error: 'Không thể lấy danh sách vai trò' });
  }
};

// Get role by ID
export const getRoleById = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const roles = await query<any[]>(
      'SELECT * FROM roles WHERE id = ?',
      [id]
    );

    if (roles.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy vai trò' });
    }

    const role = roles[0];
    const permissions = await query<any[]>(
      'SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?',
      [id]
    );

    res.json({
      ...role,
      permissions: permissions.reduce((acc, perm) => {
        acc[perm.permission_type] = perm.allowed;
        return acc;
      }, {} as Record<string, boolean>),
    });
  } catch (error: any) {
    console.error('Error fetching role:', error);
    res.status(500).json({ error: 'Không thể lấy thông tin vai trò' });
  }
};

// Create new role
export const createRole = async (req: Request, res: Response) => {
  try {
    const { name, description, permissions } = req.body;

    if (!name) {
      return res.status(400).json({ error: 'Tên vai trò là bắt buộc' });
    }

    const id = uuidv4();
    const createdAt = new Date().toISOString().slice(0, 19).replace('T', ' ');

    // Insert role
    await query(
      'INSERT INTO roles (id, name, description, created_at, updated_at) VALUES (?, ?, ?, ?, ?)',
      [id, name, description || null, createdAt, createdAt]
    );

    // Insert permissions
    if (permissions && typeof permissions === 'object') {
      const permissionTypes = [
        'view_drawing',
        'view_contract',
        'view_report',
        'view_daily_report',
      ];

      for (const permType of permissionTypes) {
        const allowed = permissions[permType] === true;
        const permId = uuidv4();
        await query(
          'INSERT INTO role_permissions (id, role_id, permission_type, allowed, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)',
          [permId, id, permType, allowed, createdAt, createdAt]
        );
      }
    }

    // Return created role with permissions
    const role = await query<any[]>(
      'SELECT * FROM roles WHERE id = ?',
      [id]
    );
    const rolePermissions = await query<any[]>(
      'SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?',
      [id]
    );

    res.status(201).json({
      ...role[0],
      permissions: rolePermissions.reduce((acc, perm) => {
        acc[perm.permission_type] = perm.allowed;
        return acc;
      }, {} as Record<string, boolean>),
    });
  } catch (error: any) {
    console.error('Error creating role:', error);
    if (error.code === 'ER_DUP_ENTRY') {
      return res.status(400).json({ error: 'Tên vai trò đã tồn tại' });
    }
    res.status(500).json({ error: 'Không thể tạo vai trò' });
  }
};

// Update role
export const updateRole = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const { name, description, permissions } = req.body;

    // Check if role exists
    const existingRoles = await query<any[]>(
      'SELECT * FROM roles WHERE id = ?',
      [id]
    );

    if (existingRoles.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy vai trò' });
    }

    const updatedAt = new Date().toISOString().slice(0, 19).replace('T', ' ');

    // Update role
    if (name !== undefined || description !== undefined) {
      await query(
        'UPDATE roles SET name = COALESCE(?, name), description = COALESCE(?, description), updated_at = ? WHERE id = ?',
        [name || null, description || null, updatedAt, id]
      );
    }

    // Update permissions
    if (permissions && typeof permissions === 'object') {
      const permissionTypes = [
        'view_drawing',
        'view_contract',
        'view_report',
        'view_daily_report',
      ];

      for (const permType of permissionTypes) {
        if (permissions[permType] !== undefined) {
          const allowed = permissions[permType] === true;
          // Check if permission exists
          const existingPerms = await query<any[]>(
            'SELECT id FROM role_permissions WHERE role_id = ? AND permission_type = ?',
            [id, permType]
          );

          if (existingPerms.length > 0) {
            // Update existing permission
            await query(
              'UPDATE role_permissions SET allowed = ?, updated_at = ? WHERE role_id = ? AND permission_type = ?',
              [allowed, updatedAt, id, permType]
            );
          } else {
            // Insert new permission
            const permId = uuidv4();
            await query(
              'INSERT INTO role_permissions (id, role_id, permission_type, allowed, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)',
              [permId, id, permType, allowed, updatedAt, updatedAt]
            );
          }
        }
      }
    }

    // Return updated role with permissions
    const role = await query<any[]>(
      'SELECT * FROM roles WHERE id = ?',
      [id]
    );
    const rolePermissions = await query<any[]>(
      'SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?',
      [id]
    );

    res.json({
      ...role[0],
      permissions: rolePermissions.reduce((acc, perm) => {
        acc[perm.permission_type] = perm.allowed;
        return acc;
      }, {} as Record<string, boolean>),
    });
  } catch (error: any) {
    console.error('Error updating role:', error);
    if (error.code === 'ER_DUP_ENTRY') {
      return res.status(400).json({ error: 'Tên vai trò đã tồn tại' });
    }
    res.status(500).json({ error: 'Không thể cập nhật vai trò' });
  }
};

// Delete role
export const deleteRole = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;

    // Check if role is being used by any user
    const usersWithRole = await query<any[]>(
      'SELECT COUNT(*) as count FROM users WHERE role = ?',
      [id]
    );

    if (usersWithRole[0]?.count > 0) {
      return res.status(400).json({
        error: 'Không thể xóa vai trò này vì đang có người dùng sử dụng',
      });
    }

    // Check if role exists
    const existingRoles = await query<any[]>(
      'SELECT * FROM roles WHERE id = ?',
      [id]
    );

    if (existingRoles.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy vai trò' });
    }

    // Delete role (permissions will be deleted by CASCADE)
    await query('DELETE FROM roles WHERE id = ?', [id]);

    res.json({ message: 'Xóa vai trò thành công' });
  } catch (error: any) {
    console.error('Error deleting role:', error);
    res.status(500).json({ error: 'Không thể xóa vai trò' });
  }
};

// Get permissions for a user's role
export const getUserPermissions = async (req: AuthRequest, res: Response) => {
  try {
    const userId = req.userId;
    if (!userId) {
      return res.status(401).json({ error: 'Unauthorized' });
    }

    // Get user's role (which is role_id)
    const users = await query<any[]>(
      'SELECT role FROM users WHERE id = ?',
      [userId]
    );

    if (users.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }

    const user = users[0];
    // role column in users table is actually role_id (UUID reference to roles table)
    const roleId = user.role;

    // Default permissions (no access)
    const defaultPermissions = {
      view_drawing: false,
      view_contract: false,
      view_report: false,
      view_daily_report: false,
    };

    if (!roleId) {
      // Admin role gets all permissions by default if no role_id is set
      if (user.role === 'admin') {
        return res.json({
          permissions: {
            view_drawing: true,
            view_contract: true,
            view_report: true,
            view_daily_report: true,
          },
        });
      }
      return res.json({ permissions: defaultPermissions });
    }

    // Get permissions for the role
    const permissions = await query<any[]>(
      'SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?',
      [roleId]
    );

    const permissionsMap = permissions.reduce((acc, perm) => {
      acc[perm.permission_type] = perm.allowed === 1 || perm.allowed === true;
      return acc;
    }, defaultPermissions as Record<string, boolean>);

    res.json({ permissions: permissionsMap });
  } catch (error: any) {
    console.error('Error fetching user permissions:', error);
    res.status(500).json({ error: 'Không thể lấy quyền của người dùng' });
  }
};

File: server.net/Controllers/RoleController.cs
using ConstructOs.Server.Config;
using ConstructOs.Server.Middleware;
using ConstructOs.Server.Utils;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Text.Json;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/roles")]
public class RoleController : ControllerBase
{
    private readonly IDatabaseService _db;

    public RoleController(IDatabaseService db)
    {
        _db = db;
    }

    [HttpGet("public")]
    public async Task<IActionResult> GetPublicRoles([FromQuery] string? search, [FromQuery] string? sortBy = "name", [FromQuery] string? sortOrder = "asc")
    {
        return await GetRoles(search, sortBy, sortOrder);
    }

    [HttpGet]
    [Authorize]
    public async Task<IActionResult> GetRoles([FromQuery] string? search, [FromQuery] string? sortBy = "name", [FromQuery] string? sortOrder = "asc")
    {
        try
        {
            var whereClause = "";
            var queryParams = new List<object>();

            if (!string.IsNullOrWhiteSpace(search))
            {
                whereClause = "WHERE name LIKE ? OR description LIKE ?";
                var searchTerm = $"%{search.Trim()}%";
                queryParams.Add(searchTerm);
                queryParams.Add(searchTerm);
            }

            var allowedSortFields = new[] { "name", "description", "created_at", "updated_at" };
            var validSortBy = allowedSortFields.Contains(sortBy) ? sortBy : "name";
            var validSortOrder = sortOrder == "desc" ? "DESC" : "ASC";

            var sql = $"SELECT * FROM roles {whereClause} ORDER BY {validSortBy} {validSortOrder}".Trim();
            var roles = await _db.QueryAsync<dynamic>(sql, queryParams.Count > 0 ? queryParams.ToArray() : null);

            var rolesWithPermissions = new List<dynamic>();
            foreach (var role in roles)
            {
                var permissions = await _db.QueryAsync<dynamic>(
                    "SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?",
                    new[] { role.id }
                );

                var permissionsDict = new Dictionary<string, bool>();
                foreach (var perm in permissions)
                {
                    permissionsDict[perm.permission_type?.ToString() ?? ""] = 
                        Convert.ToBoolean(perm.allowed ?? false);
                }

                var roleDict = (IDictionary<string, object>)role;
                roleDict["permissions"] = permissionsDict;
                rolesWithPermissions.Add(roleDict);
            }

            return Ok(rolesWithPermissions);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching roles: {ex}");
            return StatusCode(500, new { error = "Không thể lấy danh sách vai trò" });
        }
    }

    [HttpGet("{id}")]
    [Authorize]
    public async Task<IActionResult> GetRoleById(string id)
    {
        try
        {
            var roles = await _db.QueryAsync<dynamic>("SELECT * FROM roles WHERE id = ?", new[] { id });
            if (!roles.Any())
            {
                return NotFound(new { error = "Không tìm thấy vai trò" });
            }

            var role = roles.First();
            var permissions = await _db.QueryAsync<dynamic>(
                "SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?",
                new[] { id }
            );

            var permissionsDict = new Dictionary<string, bool>();
            foreach (var perm in permissions)
            {
                permissionsDict[perm.permission_type?.ToString() ?? ""] = 
                    Convert.ToBoolean(perm.allowed ?? false);
            }

            var roleDict = (IDictionary<string, object>)role;
            roleDict["permissions"] = permissionsDict;

            return Ok(roleDict);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching role: {ex}");
            return StatusCode(500, new { error = "Không thể lấy thông tin vai trò" });
        }
    }

    [HttpPost]
    [Authorize]
    public async Task<IActionResult> CreateRole([FromBody] dynamic roleData)
    {
        try
        {
            if (roleData.name == null)
            {
                return BadRequest(new { error = "Tên vai trò là bắt buộc" });
            }

            var id = Guid.NewGuid().ToString();
            var createdAt = DataHelpers.ToMySQLDateTime();

            await _db.ExecuteAsync(
                "INSERT INTO roles (id, name, description, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
                new object[] { id, roleData.name, roleData.description ?? null, createdAt, createdAt }
            );

            if (roleData.permissions != null)
            {
                var permissionTypes = new[] { "view_drawing", "view_contract", "view_report", "view_daily_report" };
                var permissions = JsonSerializer.Deserialize<Dictionary<string, bool>>(roleData.permissions.ToString() ?? "{}");

                foreach (var permType in permissionTypes)
                {
                    var allowed = permissions != null && permissions.ContainsKey(permType) && permissions[permType];
                    var permId = Guid.NewGuid().ToString();
                    await _db.ExecuteAsync(
                        "INSERT INTO role_permissions (id, role_id, permission_type, allowed, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)",
                        new object[] { permId, id, permType, allowed, createdAt, createdAt }
                    );
                }
            }

            var role = await _db.QueryAsync<dynamic>("SELECT * FROM roles WHERE id = ?", new[] { id });
            var rolePermissions = await _db.QueryAsync<dynamic>(
                "SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?",
                new[] { id }
            );

            var permissionsDict = new Dictionary<string, bool>();
            foreach (var perm in rolePermissions)
            {
                permissionsDict[perm.permission_type?.ToString() ?? ""] = 
                    Convert.ToBoolean(perm.allowed ?? false);
            }

            var roleDict = (IDictionary<string, object>)role.First();
            roleDict["permissions"] = permissionsDict;

            return StatusCode(201, roleDict);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error creating role: {ex}");
            return StatusCode(500, new { error = "Không thể tạo vai trò" });
        }
    }

    [HttpPut("{id}")]
    [Authorize]
    public async Task<IActionResult> UpdateRole(string id, [FromBody] dynamic roleData)
    {
        try
        {
            var existing = await _db.QueryAsync<dynamic>("SELECT * FROM roles WHERE id = ?", new[] { id });
            if (!existing.Any())
            {
                return NotFound(new { error = "Không tìm thấy vai trò" });
            }

            var updatedAt = DataHelpers.ToMySQLDateTime();

            if (roleData.name != null || roleData.description != null)
            {
                await _db.ExecuteAsync(
                    "UPDATE roles SET name = COALESCE(?, name), description = COALESCE(?, description), updated_at = ? WHERE id = ?",
                    new object[] { roleData.name ?? null, roleData.description ?? null, updatedAt, id }
                );
            }

            if (roleData.permissions != null)
            {
                var permissionTypes = new[] { "view_drawing", "view_contract", "view_report", "view_daily_report" };
                var permissions = JsonSerializer.Deserialize<Dictionary<string, bool>>(roleData.permissions.ToString() ?? "{}");

                foreach (var permType in permissionTypes)
                {
                    if (permissions != null && permissions.ContainsKey(permType))
                    {
                        var allowed = permissions[permType];
                        var existingPerms = await _db.QueryAsync<dynamic>(
                            "SELECT id FROM role_permissions WHERE role_id = ? AND permission_type = ?",
                            new[] { id, permType }
                        );

                        if (existingPerms.Any())
                        {
                            await _db.ExecuteAsync(
                                "UPDATE role_permissions SET allowed = ?, updated_at = ? WHERE role_id = ? AND permission_type = ?",
                                new object[] { allowed, updatedAt, id, permType }
                            );
                        }
                        else
                        {
                            var permId = Guid.NewGuid().ToString();
                            await _db.ExecuteAsync(
                                "INSERT INTO role_permissions (id, role_id, permission_type, allowed, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)",
                                new object[] { permId, id, permType, allowed, updatedAt, updatedAt }
                            );
                        }
                    }
                }
            }

            var role = await _db.QueryAsync<dynamic>("SELECT * FROM roles WHERE id = ?", new[] { id });
            var rolePermissions = await _db.QueryAsync<dynamic>(
                "SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?",
                new[] { id }
            );

            var permissionsDict = new Dictionary<string, bool>();
            foreach (var perm in rolePermissions)
            {
                permissionsDict[perm.permission_type?.ToString() ?? ""] = 
                    Convert.ToBoolean(perm.allowed ?? false);
            }

            var roleDict = (IDictionary<string, object>)role.First();
            roleDict["permissions"] = permissionsDict;

            return Ok(roleDict);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error updating role: {ex}");
            return StatusCode(500, new { error = "Không thể cập nhật vai trò" });
        }
    }

    [HttpDelete("{id}")]
    [Authorize]
    public async Task<IActionResult> DeleteRole(string id)
    {
        try
        {
            var usersWithRole = await _db.QueryAsync<dynamic>(
                "SELECT COUNT(*) as count FROM users WHERE role = ?",
                new[] { id }
            );

            if (usersWithRole.Any() && Convert.ToInt32(usersWithRole.First().count ?? 0) > 0)
            {
                return BadRequest(new { error = "Không thể xóa vai trò này vì đang có người dùng sử dụng" });
            }

            var existing = await _db.QueryAsync<dynamic>("SELECT * FROM roles WHERE id = ?", new[] { id });
            if (!existing.Any())
            {
                return NotFound(new { error = "Không tìm thấy vai trò" });
            }

            await _db.ExecuteAsync("DELETE FROM roles WHERE id = ?", new[] { id });
            return Ok(new { message = "Xóa vai trò thành công" });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error deleting role: {ex}");
            return StatusCode(500, new { error = "Không thể xóa vai trò" });
        }
    }

    [HttpGet("my-permissions")]
    [Authorize]
    public async Task<IActionResult> GetUserPermissions()
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Unauthorized" });
            }

            var users = await _db.QueryAsync<dynamic>("SELECT role FROM users WHERE id = ?", new[] { userId });
            if (!users.Any())
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            var user = users.First();
            var roleId = user.role?.ToString();

            var defaultPermissions = new Dictionary<string, bool>
            {
                { "view_drawing", false },
                { "view_contract", false },
                { "view_report", false },
                { "view_daily_report", false }
            };

            if (string.IsNullOrEmpty(roleId))
            {
                return Ok(new { permissions = defaultPermissions });
            }

            var permissions = await _db.QueryAsync<dynamic>(
                "SELECT permission_type, allowed FROM role_permissions WHERE role_id = ?",
                new[] { roleId }
            );

            var permissionsMap = defaultPermissions;
            foreach (var perm in permissions)
            {
                var permType = perm.permission_type?.ToString() ?? "";
                if (permissionsMap.ContainsKey(permType))
                {
                    permissionsMap[permType] = Convert.ToBoolean(perm.allowed ?? false);
                }
            }

            return Ok(new { permissions = permissionsMap });
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error fetching user permissions: {ex}");
            return StatusCode(500, new { error = "Không thể lấy quyền của người dùng" });
        }
    }
}

